tap_ringcentral/client.py

```diff
@@ -2,52 +2,107 @@
 import base64
 import backoff
 import time
-
+import json
 import singer
 import singer.metrics
 
+from requests.auth import HTTPBasicAuth
+from datetime import datetime, timedelta
+
 LOGGER = singer.get_logger()  # noqa
 
 
 class APIException(Exception):
     pass
 
 
+class AuthFailedException(APIException):
+    """Raised when authentication fails irrecoverably"""
+    pass
+
+
 class RingCentralClient:
 
     MAX_TRIES = 7
 
-    def __init__(self, config):
+    def __init__(self, config, config_path):
         self.config = config
+        self.config_path = config_path
         self.base_url = self.config.get('api_url')
 
-        self.refresh_token, self.access_token = self.get_authorization()
+        self.refresh_token = self.config.get('refresh_token')
+        self.access_token = self.config.get('access_token')
+    
+        self.ensure_authorization()
+
+    def write_config(self, data):
+        self.config.update(data)
+        with open(self.config_path, "w") as tap_config:
+            json.dump(self.config, tap_config, indent=2)
+
+    def is_refresh_token_expired(self):
+        expires_at = self.config.get('refresh_token_expires_at')
+        return datetime.now() >= datetime.fromisoformat(expires_at)
+
+    def is_access_token_expired(self):
+        expires_at = self.config.get('access_token_expires_at')
+        if not expires_at:
+            return True
+        return datetime.now() >= datetime.fromisoformat(expires_at)
```

Comment:
We don't need to maintain expires_at for access token. We can just refresh token whenever it expires through backoff logic.